[FEAT] Add same versions of compilers, valgrind and readline

itislu opened the pull request:
Also:
- Use same Ubuntu version as on campus for base image.
- Add more VS Code extensions.
- Let devcontainer start in `/workspaces/<project_directory>`.
- Update the docker-compose to behave same as devcontainer, except VS Code settings, extensions and current directory.
- Replace infinite sleep with better solution in docker-compose.
- Update the run script for pure Docker container with short help message and only use `docker-compose`, not also `docker build`.
- Add note in README that "The configuration is based on 42 Vienna's installed versions and tools."

itislu committed:
feat: Add same versions of compilers, valgrind and readline

itislu committed:
refactor: Handle ohmyzsh installation with devcontainer feature

itislu committed:
fix: Use `mkdir` for ensuring .ssh dir

itislu committed:
feat: Add more VS Code extensions

itislu committed:
style: Apply consistent Dockerfile formatting

Putting the `&&` in the beginning is easier to work with when modifying f.e. the last command.

itislu committed:
revert: Revert "refactor: Handle ohmyzsh installation with devcontainer feature"

Having everything important managed in the Dockerfile is better for non-devcontainer users.

This reverts commit e8ed6e8ac0ca57f7d0ceaee06e35d7d0cd24f27f.

itislu committed:
chore: Name workdir same as devcontainer would

itislu committed:
chore: Update the pure Dockerfile outside of devcontainer

itislu committed:
improve: Update the run script for pure Docker container

itislu committed:
docs: Add note that it is based on 42 Vienna

LeaYeh reviewed (changes requested):
I want to discuss with you before approving.
1. Do we really need so much vscode ext? Need to consider it might make the image fat.
2. Dose it really good to use root directly? It's also a question in my mind, so i leave the build command as old one to keep reminding me.

itislu committed:
fix: Fix git error because of directory name

git commands in `/workspace` seem to get "dubious ownership" errors, while in `/app` they don't.

itislu committed:
improve: Use current dir for workspace directory name in container

itislu committed:
fix: Replace infinite sleep with better solution

itislu committed:
improve: Give container created by docker-compose a name

itislu committed:
fix: Remove deprecated `version` key

itislu committed:
improve: Remove unnecessary VS Code extensions and add comments

itislu commented:
> 1. Do we really need so much vscode ext? Need to consider it might make the image fat.

I removed the really unnecessary ones and made short comments explaining what the others are for.

> 2. Dose it really good to use root directly? 

I don't know, I'm afraid I cannot help you with that.
